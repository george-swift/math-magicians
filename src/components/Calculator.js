import { useState } from 'react';
import { FcCalculator } from 'react-icons/fc';
import { GiPointyHat } from 'react-icons/gi';
import Display from './Display';
import ButtonPanel from './ButtonPanel';
import calculate from '../logic/calculate';

const calculatorData = {
  total: null,
  next: null,
  operation: null,
};

const Calculator = () => {
  const [data, setData] = useState(calculatorData);

  const handleClick = (buttonName) => setData((data) => calculate(data, buttonName));

  const { total, next, operation } = data;

  return (
    <div className="container wrapper">
      <div className="row w-100 me-0">
        <div className="col-md-3 mb-4 p-2">
          <h1 className="display-4">
            Let&apos;s do some math!
            <GiPointyHat />
            <FcCalculator />
          </h1>
          <ul className="mt-3">
            <li>Click the AC button to clear</li>
            <li>Click on other buttons for evaluations</li>
          </ul>
        </div>
        <div className="col-md-9 d-flex justify-content-end">
          <div className="row flex-column calculator">
            <Display operation={operation} result={next || total || '0'} />
            <ButtonPanel clickHandler={handleClick} />
          </div>
        </div>
      </div>
    </div>
  );
};

export default Calculator;
