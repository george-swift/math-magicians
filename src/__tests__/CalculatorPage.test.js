import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { render, within } from '@testing-library/react';
import Calculator from '../components/Calculator';

describe('Snapshot of Calculator Page', () => {
  test('should render page correctly', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Test calculator component', () => {
  test('Information from wireframe is on calculator page', () => {
    const { getByText } = render(<Calculator />);
    const textNode = getByText("Let's do some math!");
    expect(textNode).toBeInTheDocument();
  });

  test('Renders all the required calculator buttons in panel', () => {
    render(<Calculator />);
    const buttonPanel = document.querySelector('.calculator');
    const buttons = within(buttonPanel).getAllByRole('button');
    expect(buttons).toHaveLength(19);
  });
});
