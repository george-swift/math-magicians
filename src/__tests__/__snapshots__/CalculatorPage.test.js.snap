// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of Calculator Page should render page correctly 1`] = `
<div
  className="container wrapper"
>
  <div
    className="row w-100 me-0"
  >
    <div
      className="col-md-3 mb-4 p-2"
    >
      <h1
        className="display-4"
      >
        Let's do some math!
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416.125 42.406c-57.576.457-104.863 25.804-144.813 64.875-41.984 41.063-75 97.61-100 155.5.78 4.503 3.06 8.946 7.094 13.658 5.158 6.024 13.183 12.113 23.188 17.593 20.01 10.962 47.79 19.545 75.5 24.47 27.71 4.925 55.505 6.21 75.156 3.438 9.825-1.386 17.538-3.91 21.813-6.563 4.274-2.653 4.916-3.957 4.812-6.625l.72-.03c-3.408-42.828-6-88.797.092-131.94 2.82-19.972 7.668-39.434 15.22-57.624-31.573 31.44-62.918 65.425-86.844 94.72 35.418-70.2 86.2-121.398 141.125-168.97-11.376-1.71-22.42-2.584-33.063-2.5zM155.21 238.994c-2.033-.012-4.053-.012-6.054.006-2.453.022-4.87.065-7.28.125-23.138.575-44.227 2.91-61.876 7.188-23.532 5.703-40.466 14.888-48.78 26.03-8.317 11.144-10.08 24.667-.97 45.532 32.86 75.263 117.185 130.26 207.844 148.594 90.66 18.33 186.108.147 242.28-66.75 13.59-16.185 15.297-29.312 9.938-43.22-5.358-13.908-19.586-28.878-40.78-42.75-14.745-9.65-32.683-18.737-52.75-27.03 1.506 22.59 3.555 44.877 5.124 65.967v.219c.607 11.402-5.49 21.585-14.344 27.938-8.853 6.353-20.268 10.08-33.437 12.406-26.337 4.654-60.026 3.398-93.344-2.188-33.317-5.585-66.085-15.466-90.28-29.312-12.097-6.923-22.145-14.85-28.875-24.47-6.73-9.617-9.76-21.554-6.594-33.374l.095-.375.125-.374c7.637-21.206 16.308-42.79 26.094-64.094-2.053-.032-4.1-.056-6.133-.068zm6.634 46.662c-3.08 7.8-6.017 15.596-8.813 23.344-1.595 6.246-.4 11.407 3.907 17.563 4.374 6.25 12.28 12.923 22.844 18.968 21.128 12.09 52.4 21.78 84.095 27.095 31.694 5.314 64.016 6.28 87 2.22 11.492-2.032 20.53-5.42 25.78-9.19 5.25-3.766 6.864-6.726 6.595-11.78-.517-6.93-1.088-14.027-1.688-21.25-7.448 4.03-16.47 6.367-26.718 7.813-22.732 3.206-51.79 1.665-81.03-3.532-29.242-5.196-58.5-14.055-81.22-26.5-11.36-6.222-21.122-13.34-28.375-21.812-.825-.962-1.62-1.933-2.376-2.938z"
            fillRule="evenodd"
          />
        </svg>
        <svg
          enableBackground="new 0 0 48 48"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          version="1"
          viewBox="0 0 48 48"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M40,16H8v24c0,2.2,1.8,4,4,4h24c2.2,0,4-1.8,4-4V16z"
            fill="#616161"
          />
          <path
            d="M36,4H12C9.8,4,8,5.8,8,8v9h32V8C40,5.8,38.2,4,36,4z"
            fill="#424242"
          />
          <path
            d="M36,14H12c-0.6,0-1-0.4-1-1V8c0-0.6,0.4-1,1-1h24c0.6,0,1,0.4,1,1v5C37,13.6,36.6,14,36,14z"
            fill="#9CCC65"
          />
          <g
            fill="#33691E"
          >
            <rect
              height="2"
              width="2"
              x="33"
              y="10"
            />
            <rect
              height="2"
              width="2"
              x="29"
              y="10"
            />
          </g>
          <path
            d="M36,23h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C37,22.6,36.6,23,36,23z"
            fill="#FF5252"
          />
          <g
            fill="#E0E0E0"
          >
            <path
              d="M15,23h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C16,22.6,15.6,23,15,23z"
            />
            <path
              d="M22,23h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C23,22.6,22.6,23,22,23z"
            />
            <path
              d="M29,23h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C30,22.6,29.6,23,29,23z"
            />
            <path
              d="M15,29h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C16,28.6,15.6,29,15,29z"
            />
            <path
              d="M22,29h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C23,28.6,22.6,29,22,29z"
            />
            <path
              d="M29,29h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C30,28.6,29.6,29,29,29z"
            />
            <path
              d="M15,35h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C16,34.6,15.6,35,15,35z"
            />
            <path
              d="M22,35h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C23,34.6,22.6,35,22,35z"
            />
            <path
              d="M29,35h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C30,34.6,29.6,35,29,35z"
            />
            <path
              d="M15,41h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C16,40.6,15.6,41,15,41z"
            />
            <path
              d="M22,41h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C23,40.6,22.6,41,22,41z"
            />
            <path
              d="M29,41h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C30,40.6,29.6,41,29,41z"
            />
          </g>
          <g
            fill="#BDBDBD"
          >
            <path
              d="M36,29h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C37,28.6,36.6,29,36,29z"
            />
            <path
              d="M36,35h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C37,34.6,36.6,35,36,35z"
            />
            <path
              d="M36,41h-3c-0.6,0-1-0.4-1-1v-2c0-0.6,0.4-1,1-1h3c0.6,0,1,0.4,1,1v2C37,40.6,36.6,41,36,41z"
            />
          </g>
        </svg>
      </h1>
      <ul
        className="mt-3"
      >
        <li>
          Click the AC button to clear
        </li>
        <li>
          Click on other buttons for evaluations
        </li>
      </ul>
    </div>
    <div
      className="col-md-9 d-flex justify-content-end"
    >
      <div
        className="row flex-column calculator"
      >
        <div
          className="col-12 h-25 text-white"
        >
          <div
            className="problem"
          />
          <div
            className="result"
          >
            0
          </div>
        </div>
        <div
          className="col-12 h-75"
        >
          <div
            className="group-one"
          >
            <button
              className="clear"
              id="allclear"
              name="allclear"
              onClick={[Function]}
              type="button"
              value="AC"
            >
              AC
            </button>
            <button
              className="operator"
              id="plusMinus"
              name="plusMinus"
              onClick={[Function]}
              type="button"
              value="+/-"
            >
              +/-
            </button>
            <button
              className="operator"
              id="percentage"
              name="percentage"
              onClick={[Function]}
              type="button"
              value="%"
            >
              %
            </button>
            <button
              className="operator"
              id="division"
              name="division"
              onClick={[Function]}
              type="button"
              value="/"
            >
              /
            </button>
          </div>
          <div
            className="group-two"
          >
            <button
              className="number"
              id="number-7"
              name="number-7"
              onClick={[Function]}
              type="button"
              value={7}
            >
              7
            </button>
            <button
              className="number"
              id="number-8"
              name="number-8"
              onClick={[Function]}
              type="button"
              value={8}
            >
              8
            </button>
            <button
              className="number"
              id="number-9"
              name="number-9"
              onClick={[Function]}
              type="button"
              value={9}
            >
              9
            </button>
            <button
              className="operator"
              id="multiply"
              name="multiply"
              onClick={[Function]}
              type="button"
              value="x"
            >
              x
            </button>
          </div>
          <div
            className="group-three"
          >
            <button
              className="number"
              id="number-4"
              name="number-4"
              onClick={[Function]}
              type="button"
              value={4}
            >
              4
            </button>
            <button
              className="number"
              id="number-5"
              name="number-5"
              onClick={[Function]}
              type="button"
              value={5}
            >
              5
            </button>
            <button
              className="number"
              id="number-6"
              name="number-6"
              onClick={[Function]}
              type="button"
              value={6}
            >
              6
            </button>
            <button
              className="operator"
              id="subtract"
              name="subtract"
              onClick={[Function]}
              type="button"
              value="-"
            >
              -
            </button>
          </div>
          <div
            className="group-four"
          >
            <button
              className="number"
              id="number-1"
              name="number-1"
              onClick={[Function]}
              type="button"
              value={1}
            >
              1
            </button>
            <button
              className="number"
              id="number-2"
              name="number-2"
              onClick={[Function]}
              type="button"
              value={2}
            >
              2
            </button>
            <button
              className="number"
              id="number-3"
              name="number-3"
              onClick={[Function]}
              type="button"
              value={3}
            >
              3
            </button>
            <button
              className="operator"
              id="plus"
              name="plus"
              onClick={[Function]}
              type="button"
              value="+"
            >
              +
            </button>
          </div>
          <div
            className="group-five"
          >
            <button
              className="number"
              id="number-0"
              name="number-0"
              onClick={[Function]}
              type="button"
              value={0}
            >
              0
            </button>
            <button
              className="decimal"
              id="decimal"
              name="decimal"
              onClick={[Function]}
              type="button"
              value="."
            >
              .
            </button>
            <button
              className="operator"
              id="equal"
              name="equal"
              onClick={[Function]}
              type="button"
              value="="
            >
              =
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
